File: android/app/src/main/java/com/fieldrepository/app/data/Offline.kt
package com.fieldrepository.app.data

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkCapabilities
import android.net.Uri
import android.provider.OpenableColumns
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.decodeFromString
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File
import java.io.FileOutputStream
import java.io.IOException
import java.time.Instant
import java.util.UUID
import java.util.concurrent.atomic.AtomicReference

/**
 * Offline outbox. When the device has no connection, a new record (and copies of its captured media)
 * is persisted to local app storage instead of being sent to the server. A sync pass (on reconnect or
 * app start) replays every queued entry — creating the record, then uploading its media — and only
 * then removes the local copy, so field data is never lost in transit.
 */

/** A captured file copied into local app storage, plus the params needed to upload it later. */
@Serializable
data class PendingMedia(
    val localPath: String,
    val originalFilename: String,
    val mimeType: String,
    val mediaType: String,
    val caption: String? = null,
    val recordName: String? = null,
    val customSegment: String? = null,
    val overrideBaseName: String? = null,
    val batchIndex: Int = 1,
    val stageStep: Int? = null,
    val processing: List<String>? = null,
    // Override the link target type (e.g. "processstep" for a process step's media). Null = the entry's
    // own type. `stepIndex` (process only) selects which created step's server id to attach to on sync.
    val linkedType: String? = null,
    val stepIndex: Int? = null
)

/** One queued create: the record type, its serialized create request, and the media to attach after. */
@Serializable
data class PendingEntry(
    val id: String,
    val type: String,
    val payloadJson: String,
    val label: String,
    val media: List<PendingMedia> = emptyList(),
    val createdAt: String,
    /**
     * REPLAY PROGRESS, written back as each step lands. All defaulted, so an entry queued by an
     * older build decodes into "nothing has happened yet" and replays exactly as it used to.
     *
     * [createdId] is the load-bearing one: non-null means the record IS on the server, and re-sending
     * the body would make a SECOND one. A replay that finds it set skips the create and goes straight
     * to the media. Without it, every media failure re-created the record on the next pass — once per
     * pass, for as long as the signal stayed bad, which is the whole reason the entry is here.
     */
    val createdId: String? = null,
    /** A created process's step ids in submit order, for the media that attaches to a `processstep`. */
    val createdStepIds: List<String> = emptyList(),
    /** Indices into [media] whose bytes are already on the server. Never uploaded twice. */
    val uploadedMedia: List<Int> = emptyList(),
    /**
     * The server's final answer, when it refused this entry for a reason no retry will change.
     *
     * The entry and its copied files STAY on the device — being refused is not a reason to destroy
     * the only copy of a day's fieldwork — and the sync pass steps over it instead of stopping, so
     * one unacceptable record cannot strand the records queued behind it.
     */
    val failure: String? = null,
    /** When [failure] was recorded, ISO-8601. */
    val failedAt: String? = null
)

/** One captured media item to stage for an offline entry (input form for staging). */
data class OfflineMediaSpec(
    val uri: Uri,
    val caption: String? = null,
    val recordName: String? = null,
    val customSegment: String? = null,
    val overrideBaseName: String? = null,
    val batchIndex: Int = 1,
    val stageStep: Int? = null,
    val processing: List<String>? = null,
    val linkedType: String? = null,
    val stepIndex: Int? = null
)

/** Live connectivity check (validated internet, not just an attached interface). */
object ConnectivityObserver {
    fun isOnline(context: Context): Boolean {
        val cm = context.getSystemService(ConnectivityManager::class.java) ?: return false
        val network = cm.activeNetwork ?: return false
        val caps = cm.getNetworkCapabilities(network) ?: return false
        return caps.hasCapability(NetworkCapabilities.NET_CAPABILITY_INTERNET) &&
            caps.hasCapability(NetworkCapabilities.NET_CAPABILITY_VALIDATED)
    }
}

object OfflineOutbox {
    private val json = Json { ignoreUnknownKeys = true; encodeDefaults = true }
    private val mutex = Mutex()

    /**
     * Something the researcher has to be told about the queue itself, waiting to be said once.
     *
     * Only [read] raises one, and only when a queue file would not parse — a silent failure with no
     * other symptom, because the queue simply comes back empty and the banner counting entries drops
     * to zero. Taken (and cleared) by the sync pass, which has a Context and can say it out loud.
     */
    private val alert = AtomicReference<String?>(null)

    private fun dir(context: Context): File = File(context.filesDir, "outbox").apply { mkdirs() }
    private fun mediaDir(context: Context): File = File(dir(context), "media").apply { mkdirs() }
    private fun queueFile(context: Context): File = File(dir(context), "queue.json")

    /** The one pending queue-level alert, cleared as it is taken so it is reported exactly once. */
    fun takeAlert(): String? = alert.getAndSet(null)

    /**
     * Read the queue, or quarantine it and say so.
     *
     * A FILE THAT WILL NOT PARSE IS NEVER TREATED AS AN EMPTY QUEUE. It used to be — a bare
     * `getOrDefault(emptyList())` — and the next [write] then overwrote the damage with whatever was
     * left, so a queue full of unsent fieldwork became zero entries, with no error, no trace and no
     * way back. The bytes are moved aside under a name a support request can ask for instead, the
     * researcher is told the count they can see is not the whole story, and the app carries on able
     * to queue new work (refusing to would turn one bad file into a device that cannot save at all).
     *
     * A ZERO-LENGTH file counts as damage too: [write] never produces one — an empty queue is the two
     * bytes `[]` — so a truncated file is the fingerprint of the non-atomic write this class used to
     * do, killed between truncating the old contents and writing the new ones.
     */
    private fun read(context: Context): List<PendingEntry> {
        val file = queueFile(context)
        if (!file.exists()) return emptyList()
        val text = runCatching { file.readText() }.getOrElse { error ->
            quarantine(file, "could not be read (${error.javaClass.simpleName})")
            return emptyList()
        }
        if (text.isBlank()) {
            quarantine(file, "was empty, which means a save was interrupted part-way through")
            return emptyList()
        }
        return runCatching { json.decodeFromString<List<PendingEntry>>(text) }.getOrElse { error ->
            quarantine(file, "could not be understood (${error.javaClass.simpleName})")
            emptyList()
        }
    }

    /** Move a damaged queue file out of the way, keeping every byte, and raise the alert. */
    private fun quarantine(file: File, reason: String) {
        val kept = File(file.parentFile, "queue.damaged-${System.currentTimeMillis()}.json")
        val moved = runCatching { file.renameTo(kept) }.getOrDefault(false)
        alert.set(
            if (moved) {
                "The list of records saved on this device $reason. Nothing has been deleted — it has been " +
                    "kept as ${kept.name} in the app's storage. Please report this before uninstalling the app."
            } else {
                "The list of records saved on this device $reason, and could not be set aside. Please report " +
                    "this before uninstalling the app."
            }
        )
    }

    /**
     * Replace the queue file with [entries] as one indivisible step: write a temp file, flush it to
     * the actual storage, then rename it over the real one.
     *
     * This used to be a plain `writeText`, which truncates the file and then writes — so a process
     * killed in that window (a low-memory kill, a flat battery, the researcher swiping the app away
     * while it drained) left a truncated or half-written queue, and the read above turned that into
     * an empty one. Every record queued on that device was gone, silently. A rename within one
     * directory is atomic: either the old file is there or the whole new one is, never a mixture.
     * The `sync()` matters as much as the rename — a rename whose bytes are still only in the page
     * cache publishes an empty file after a power loss.
     */
    private fun write(context: Context, entries: List<PendingEntry>) {
        val target = queueFile(context)
        val temp = File(target.parentFile, "${target.name}.writing")
        val bytes = json.encodeToString(entries).toByteArray()
        try {
            FileOutputStream(temp).use { out ->
                out.write(bytes)
                out.flush()
                out.fd.sync()
            }
            if (!temp.renameTo(target)) throw IOException("Unable to replace ${target.name}")
        } catch (e: Throwable) {
            runCatching { temp.delete() }
            throw e
        }
    }

    // Every read and write below runs on the IO dispatcher. These are called from `LaunchedEffect`s
    // and save handlers that start on the main dispatcher, and reading the whole queue — every
    // record, every media descriptor — off flash on the main thread is how a form janks on the frame
    // it saves. The lock is taken INSIDE the dispatcher switch so a waiter never holds a main-thread
    // frame while another coroutine finishes its file IO.

    suspend fun all(context: Context): List<PendingEntry> =
        withContext(Dispatchers.IO) { mutex.withLock { read(context) } }

    suspend fun count(context: Context): Int =
        withContext(Dispatchers.IO) { mutex.withLock { read(context).size } }

    suspend fun enqueue(context: Context, entry: PendingEntry) = withContext(Dispatchers.IO) {
        mutex.withLock { write(context, read(context) + entry) }
    }

    /**
     * Write [entry] back over the queued copy carrying the same id.
     *
     * Read-modify-write under the lock and addressed by id, never by writing a snapshot back
     * wholesale: a sync pass replays a snapshot taken at its start while the researcher may still be
     * saving new records into the same file, and a wholesale write would delete them. An entry that
     * is no longer queued is left alone rather than resurrected.
     */
    suspend fun update(context: Context, entry: PendingEntry) = withContext(Dispatchers.IO) {
        mutex.withLock {
            val current = read(context)
            if (current.any { it.id == entry.id }) {
                write(context, current.map { if (it.id == entry.id) entry else it })
            }
        }
    }

    /**
     * Record why the server will never accept this entry. The entry and its files are KEPT: a refusal
     * is a reason to stop retrying and tell someone, not a reason to destroy the only copy.
     */
    suspend fun markFailure(context: Context, entryId: String, reason: String) = withContext(Dispatchers.IO) {
        mutex.withLock {
            val current = read(context)
            if (current.any { it.id == entryId }) {
                write(
                    context,
                    current.map {
                        if (it.id == entryId) it.copy(failure = reason, failedAt = Instant.now().toString()) else it
                    }
                )
            }
        }
    }

    /** Every entry the server has refused, for a surface that wants to show or discard them. */
    suspend fun failed(context: Context): List<PendingEntry> =
        withContext(Dispatchers.IO) { mutex.withLock { read(context).filter { it.failure != null } } }

    /** Remove a synced entry and delete the local media copies it owned. */
    suspend fun remove(context: Context, entry: PendingEntry) = withContext(Dispatchers.IO) {
        mutex.withLock { write(context, read(context).filterNot { it.id == entry.id }) }
        entry.media.forEach { runCatching { File(it.localPath).delete() } }
    }

    /** Copy a captured content Uri into local app storage so it survives offline until uploaded. */
    fun stageMedia(
        context: Context,
        uri: Uri,
        caption: String?,
        recordName: String?,
        customSegment: String?,
        overrideBaseName: String?,
        batchIndex: Int,
        processing: List<String>?,
        stageStep: Int? = null,
        linkedType: String? = null,
        stepIndex: Int? = null
    ): PendingMedia {
        val mimeType = context.contentResolver.getType(uri) ?: "application/octet-stream"
        val originalName = displayName(context, uri) ?: "field-media-${System.currentTimeMillis()}"
        val extension = originalName.substringAfterLast('.', "").takeIf { it.isNotBlank() }
        val target = File(mediaDir(context), UUID.randomUUID().toString() + (extension?.let { ".$it" } ?: ""))
        context.contentResolver.openInputStream(uri)?.use { input ->
            FileOutputStream(target).use { output -> input.copyTo(output, 64 * 1024) }
        } ?: throw IllegalStateException("Unable to read the captured media for offline storage")
        return PendingMedia(
            localPath = target.absolutePath,
            originalFilename = originalName,
            mimeType = mimeType,
            mediaType = inferMediaType(mimeType),
            caption = caption,
            recordName = recordName,
            customSegment = customSegment,
            overrideBaseName = overrideBaseName,
            batchIndex = batchIndex,
            stageStep = stageStep,
            processing = processing,
            linkedType = linkedType,
            stepIndex = stepIndex
        )
    }

    private fun displayName(context: Context, uri: Uri): String? {
        context.contentResolver.query(uri, arrayOf(OpenableColumns.DISPLAY_NAME), null, null, null)?.use { cursor ->
            val index = cursor.getColumnIndex(OpenableColumns.DISPLAY_NAME)
            if (index >= 0 && cursor.moveToFirst()) return cursor.getString(index)
        }
        return uri.lastPathSegment
    }

    private fun inferMediaType(mimeType: String): String = when {
        mimeType.startsWith("image/") -> "IMAGE"
        mimeType.startsWith("video/") -> "VIDEO"
        mimeType.startsWith("audio/") -> "AUDIO"
        mimeType == "application/pdf" -> "PDF"
        else -> "DOCUMENT"
    }
}

File: frontend/lib/offline.ts
"use client";

/**
 * The offline outbox — the web's port of the Android app's `OfflineOutbox`.
 *
 * A researcher in a village with no signal must be able to finish the interview they are in the
 * middle of. On Android they always could: a save with no connection goes into a local outbox and
 * drains when the network returns. On the web it used to fail at the Save button, and the only
 * honest thing the UI could do was warn them (see `components/dialogs/OfflineDialog`). This is the
 * outbox that warning was apologising for.
 *
 * WHAT IS STORED. One entry per attempted save: the record request (endpoint, method, JSON body)
 * and the files that were attached to it. Files go in as `File` objects — IndexedDB stores them by
 * structured clone, so the bytes, the name and the MIME type all survive a browser restart, which a
 * blob: URL or an in-memory array would not. This matters: the attachments are usually the part
 * that cannot be recreated, because the artisan has gone home.
 *
 * HOW IT DRAINS, AND HOW IT DIFFERS FROM ANDROID. `syncOutbox` replays entries oldest first: create
 * the record, then upload its media against the new id, then delete the entry. The Android outbox
 * stops at the first failure, which is right for a connection that dropped again — but wrong for a
 * request the server will never accept: one 422 at the head of the queue blocks every entry behind
 * it forever, and nothing tells the user why. So the failure is triaged here:
 *
 *   - no connection / 5xx / timeout  → transient. Stop the pass, keep everything queued, try again
 *                                      on the next `online` event. Nothing is lost.
 *   - 4xx (validation, permission)   → permanent. Mark THAT entry with the server's reason, leave
 *                                      it in the outbox for the user to see and discard, and carry
 *                                      on to the next one. One bad record cannot strand the others.
 *
 * A REPLAY IS RESUMABLE, because "create then upload" is two steps and only the first is cheap to
 * repeat — repeating it makes a second record. So each step is written back to the entry the moment
 * it lands (`created`, `createdId`, `uploadedBatches`), and a pass that dies half way through picks
 * up at the media instead of starting the record again. Without that the outbox duplicated every
 * record whose media upload was interrupted, once per sync pass, for as long as the signal stayed
 * bad — the failure mode with the worst timing possible, since a bad signal is why the entry is here.
 *
 * NOTHING IS EVER DELETED BECAUSE THE SERVER SAID 409. This module used to read a 409 as "the create
 * already landed and we simply lost the response", and drop the entry and its files as sent. No
 * endpoint in this API means that: a 409 from /artisans is a clashing Aadhaar, from /crafts a craft
 * of that name, from /questionnaire/interviews the same artisan set already interviewed. So the one
 * answer that means "someone else's record collides with yours" was destroying the record AND the
 * photographs and reporting success. A 409 is now surfaced as a conflict for the researcher to
 * resolve, with everything kept. The lost-response case it was aiming at is covered properly by
 * `created` above, which knows rather than guesses.
 */

import { ApiError, apiFetch } from "@/lib/api";
import { uploadMediaBatch } from "@/lib/media";

const DB_NAME = "field-repo-outbox";
const DB_VERSION = 1;
const STORE = "entries";

/**
 * One media batch of a queued save: everything `uploadMediaBatch` will need on replay.
 *
 * A LIST of these, not one, because the forms do not attach media in a single lump: a product
 * queues its two measurement-grid photos (each with its own caption naming the dimension) beside
 * the general field media, and a tool adds its numbered process-stage captures on top. Flattening
 * them into one batch would put every file under one caption, and the caption is the only thing
 * that says which photo is the height grid.
 */
export type OutboxMediaBatch = {
  files: File[];
  linkedRecordType: string;
  caption?: string;
  location?: unknown;
  recordedAt?: string | null;
  recordedTimezone?: string | null;
  extraMetadata?: Record<string, unknown>;
  transcribeAudio?: boolean;
  /**
   * Link this batch to the Nth CHILD of the created record instead of the record itself — the
   * process form's per-step media, whose `processstep` ids do not exist until the server has made
   * them. Resolved on replay from the create response's `steps[]`, in the order they were sent.
   */
  stepIndex?: number;
};

export type OutboxEntry = {
  id?: number;
  /** Human label for the banner — "Artisan · Giriraj Prasad", not an endpoint. */
  label: string;
  createdAt: number;
  endpoint: string;
  method: "POST" | "PATCH";
  /** Serialised at queue time so a later schema change cannot alter what the user actually saved. */
  body: string;
  media: OutboxMediaBatch[];
  attempts: number;
  /** Set when the server rejected this permanently; the entry stays for the user to read and discard. */
  failure: string | null;
  /**
   * Replay progress. All optional: entries written before this existed simply have none, and start
   * their replay from the top exactly as they used to.
   *
   * `created` is the load-bearing one — true means the record IS on the server and re-sending the
   * body would make a second one, so the replay must skip straight to the media. The rest say what
   * the media may link to and which batches are already up there.
   */
  created?: boolean;
  createdId?: string | null;
  /** Ids of the created record's children, for batches that address a `processstep` by index. */
  createdStepIds?: string[];
  /** Indices into `media` whose files have already been sent; never sent twice. */
  uploadedBatches?: number[];
};

// ---------------------------------------------------------------------------
// IndexedDB plumbing
// ---------------------------------------------------------------------------

let dbPromise: Promise<IDBDatabase> | null = null;

function openDb(): Promise<IDBDatabase> {
  if (typeof indexedDB === "undefined") return Promise.reject(new Error("IndexedDB unavailable"));
  if (!dbPromise) {
    dbPromise = new Promise((resolve, reject) => {
      const request = indexedDB.open(DB_NAME, DB_VERSION);
      request.onupgradeneeded = () => {
        const db = request.result;
        if (!db.objectStoreNames.contains(STORE)) {
          db.createObjectStore(STORE, { keyPath: "id", autoIncrement: true });
        }
      };
      request.onsuccess = () => resolve(request.result);
      request.onerror = () => reject(request.error ?? new Error("Cannot open the offline outbox"));
    });
    // A failed open must not be cached forever — a private-mode tab that later allows storage
    // should get a working outbox rather than the first rejection for the rest of the session.
    dbPromise.catch(() => {
      dbPromise = null;
    });
  }
  return dbPromise;
}

function tx<T>(mode: IDBTransactionMode, run: (store: IDBObjectStore) => IDBRequest<T>): Promise<T> {
  return openDb().then(
    (db) =>
      new Promise<T>((resolve, reject) => {
        const transaction = db.transaction(STORE, mode);
        const request = run(transaction.objectStore(STORE));
        request.onsuccess = () => resolve(request.result);
        request.onerror = () => reject(request.error ?? new Error("Offline outbox write failed"));
      })
  );
}

// ---------------------------------------------------------------------------
// Subscription — one source of truth for every banner on the page
// ---------------------------------------------------------------------------

const listeners = new Set<() => void>();
let cache: OutboxEntry[] = [];

function publish() {
  listeners.forEach((listener) => listener());
}

export function subscribeOutbox(listener: () => void): () => void {
  listeners.add(listener);
  return () => listeners.delete(listener);
}

/** Synchronous snapshot for `useSyncExternalStore`; refreshed by {@link refreshOutbox}. */
export function getOutboxSnapshot(): OutboxEntry[] {
  return cache;
}

/** Server render has no IndexedDB — a stable empty array keeps the store from looping. */
const SERVER_SNAPSHOT: OutboxEntry[] = [];
export function getServerOutboxSnapshot(): OutboxEntry[] {
  return SERVER_SNAPSHOT;
}

export async function refreshOutbox(): Promise<OutboxEntry[]> {
  try {
    const rows = await tx<OutboxEntry[]>("readonly", (store) => store.getAll() as IDBRequest<OutboxEntry[]>);
    cache = rows.sort((a, b) => a.createdAt - b.createdAt);
  } catch {
    cache = [];
  }
  publish();
  return cache;
}

// ---------------------------------------------------------------------------
// Queue / discard
// ---------------------------------------------------------------------------

export async function queueOffline(entry: Omit<OutboxEntry, "id" | "createdAt" | "attempts" | "failure">): Promise<void> {
  await tx("readwrite", (store) =>
    store.add({ ...entry, createdAt: Date.now(), attempts: 0, failure: null } as OutboxEntry)
  );
  await refreshOutbox();
}

export async function discardOutboxEntry(id: number): Promise<void> {
  await tx("readwrite", (store) => store.delete(id));
  await refreshOutbox();
}

async function markFailure(entry: OutboxEntry, failure: string): Promise<void> {
  await tx("readwrite", (store) => store.put({ ...entry, attempts: entry.attempts + 1, failure }));
}

/**
 * Write a replay's progress back to the entry mid-pass.
 *
 * Called after the create lands and after each media batch goes up, so the durable record of what
 * has already happened is never behind what actually happened. The window between the server
 * committing and this returning is the only place a duplicate can still be born; a few milliseconds
 * of IndexedDB is as small as that window gets without idempotency keys on the API.
 */
async function persistProgress(entry: OutboxEntry): Promise<void> {
  await tx("readwrite", (store) => store.put(entry));
}

/** Files across every batch of one entry — what the researcher stands to lose, in one number. */
function pendingFileCount(entry: OutboxEntry): number {
  return entry.media.reduce((sum, batch) => sum + batch.files.length, 0);
}

// ---------------------------------------------------------------------------
// Drain
// ---------------------------------------------------------------------------

/** A fetch that never reached a server: still offline, so stop the pass and keep everything. */
function isTransient(error: unknown): boolean {
  if (error instanceof ApiError) return error.status >= 500 || error.status === 408 || error.status === 429;
  return true; // TypeError from fetch, an abort, a DNS failure — all "try again later".
}

export type SyncResult = { synced: number; failed: number; remaining: number; stoppedOffline: boolean };

let syncing: Promise<SyncResult> | null = null;

/**
 * Replay the outbox oldest first. Concurrent calls (the `online` event and a "Sync now" click
 * landing together) share one pass — replaying an entry twice would create the record twice.
 */
export function syncOutbox(): Promise<SyncResult> {
  if (syncing) return syncing;
  syncing = runSync().finally(() => {
    syncing = null;
  });
  return syncing;
}

async function runSync(): Promise<SyncResult> {
  const entries = await refreshOutbox();
  let synced = 0;
  let failed = 0;
  let stoppedOffline = false;

  for (const entry of entries) {
    if (entry.failure) continue; // Already triaged as permanent; waiting on the user, not the network.
    // Everything this pass achieves is recorded on `progress` and written through as it happens, so
    // that an interruption resumes rather than restarts. See the resumability note at the top.
    const progress: OutboxEntry = { ...entry };
    try {
      if (!progress.created) {
        try {
          const saved = await apiFetch<{ id: string; steps?: Array<{ id: string }> }>(progress.endpoint, {
            method: progress.method,
            body: progress.body
          });
          // Every endpoint the outbox replays answers with the saved record, so a 2xx carrying no id
          // did not come from one — a captive portal answering the POST with its own 200 page is the
          // field case, and this app already knows they exist (see the "Sync now" note in
          // OutboxBanner). Taking it as success set `createdId` to null, which skipped the media loop
          // in silence and then discarded the entry: no record written, the photographs deleted, and
          // "sent" reported. An answer we cannot read is not a save — keep the entry and say so.
          const createdId = typeof saved?.id === "string" && saved.id ? saved.id : null;
          if (!createdId) {
            const files = pendingFileCount(progress);
            await markFailure(
              progress,
              "The server accepted this but did not say what it saved, so it cannot be confirmed or its files " +
                `attached. This entry${files ? ` and its ${files} file(s)` : ""} are still on this device. If you were ` +
                "on a wi-fi network that asks you to sign in, connect properly and check whether the record arrived " +
                "before discarding this."
            );
            failed += 1;
            continue;
          }
          progress.created = true;
          progress.createdId = createdId;
          progress.createdStepIds = (saved?.steps ?? []).map((step) => step.id);
        } catch (error) {
          // A 409 is a COLLISION WITH SOMEONE ELSE'S RECORD — a clashing Aadhaar, a craft already
          // named that — not an echo of our own earlier create. Deleting the entry here is what used
          // to throw away the queued record and its photographs and then report them as sent. The
          // researcher is the only one who can tell whether the existing record is theirs, so hand
          // it to them with everything intact.
          if (error instanceof ApiError && error.status === 409) {
            const files = pendingFileCount(progress);
            await markFailure(
              progress,
              `The server refused this as a duplicate. ${error.message} Nothing has been sent and nothing has been ` +
                `thrown away — this entry${files ? ` and its ${files} file(s)` : ""} are still on this device. Open the ` +
                "record it clashes with, carry across anything it is missing, then discard this entry."
            );
            failed += 1;
            continue;
          }
          throw error;
        }
        // Persisted BEFORE a single byte of media moves: from here on the record exists, and a pass
        // that dies during the upload must come back to the media, never to the create.
        await persistProgress(progress);
      }

      const mediaFailed: Array<{ name: string }> = [];
      const uploaded = new Set(progress.uploadedBatches ?? []);
      if (progress.createdId) {
        for (const [index, batch] of progress.media.entries()) {
          if (uploaded.has(index) || !batch.files.length) continue;
          // A step batch has to wait for the server to mint the step; if the create came back with
          // fewer steps than were queued, say so rather than silently attaching to the wrong one.
          const linkedRecordId =
            batch.stepIndex === undefined ? progress.createdId : (progress.createdStepIds?.[batch.stepIndex] ?? null);
          if (!linkedRecordId) {
            mediaFailed.push(...batch.files.map((file) => ({ name: file.name })));
            continue;
          }
          const result = await uploadMediaBatch({
            files: batch.files,
            linkedRecordType: batch.linkedRecordType,
            linkedRecordId,
            caption: batch.caption,
            location: batch.location,
            recordedAt: batch.recordedAt ?? undefined,
            recordedTimezone: batch.recordedTimezone ?? undefined,
            extraMetadata: batch.extraMetadata,
            transcribeAudio: batch.transcribeAudio ?? true
          });
          mediaFailed.push(...result.failed);
          // Returning at all means this batch got as far as it ever will: `uploadMediaBatch` only
          // throws when NOTHING in it landed, so a throw leaves the batch unrecorded and retriable
          // while a return — even a partly failed one — must never be replayed into duplicate files.
          uploaded.add(index);
          progress.uploadedBatches = Array.from(uploaded);
          await persistProgress(progress);
        }
      }
      // The record exists now, so the entry cannot simply be retried — re-running it would create a
      // second record just to have another go at a file. Report which files did not make it instead
      // of dropping them silently.
      if (mediaFailed.length) {
        await markFailure(
          progress,
          `Saved, but ${mediaFailed.length} file(s) did not upload: ${mediaFailed.map((f) => f.name).join(", ")}. ` +
            "Re-attach them on the record."
        );
        failed += 1;
        await refreshOutbox();
        continue;
      }

      await discardOutboxEntry(progress.id!);
      synced += 1;
    } catch (error) {
      if (isTransient(error)) {
        stoppedOffline = true;
        break; // Still offline (or the API is down) — everything behind this stays queued.
      }
      await markFailure(progress, error instanceof Error ? error.message : "The server rejected this entry.");
      failed += 1;
    }
  }

  const remaining = (await refreshOutbox()).length;
  return { synced, failed, remaining, stoppedOffline };
}

// ---------------------------------------------------------------------------
// The one call a form makes
// ---------------------------------------------------------------------------

export type SaveOutcome<T> = { queued: true } | { queued: false; saved: T };

/**
 * Save a record, or queue it if this device has no connection.
 *
 * Deliberately does NOT upload the media when online: the caller keeps its own `uploadMediaBatch`
 * call so its progress bar, per-file retry and eager-staging claim all behave exactly as before.
 * `media` is taken here only so that a QUEUED save can carry the files into IndexedDB with it.
 */
export async function saveOrQueue<T extends { id: string }>({
  label,
  endpoint,
  method,
  body,
  media
}: {
  label: string;
  endpoint: string;
  method: "POST" | "PATCH";
  body: unknown;
  /** Batches to persist WITH a queued save. Ignored when the save goes through online. */
  media?: OutboxMediaBatch[];
}): Promise<SaveOutcome<T>> {
  const payload = JSON.stringify(body);
  const queue = async () => {
    await queueOffline({ label, endpoint, method, body: payload, media: (media ?? []).filter((batch) => batch.files.length) });
    return { queued: true } as const;
  };

  // Known-offline: do not burn a request (and a 30s timeout) to learn what the browser already knows.
  if (typeof navigator !== "undefined" && navigator.onLine === false) return queue();

  try {
    return { queued: false, saved: await apiFetch<T>(endpoint, { method, body: payload }) };
  } catch (error) {
    // Only a request that never reached the server may be queued. A 4xx means the server saw it and
    // said no; queueing that would replay a rejection forever and hide the real problem from the user.
    if (isTransient(error) && !(error instanceof ApiError)) return queue();
    throw error;
  }
}
